modules/reindex/src/test/java/org/elasticsearch/index/reindex/RetryTests.java

```diff
@@ -20,136 +20,159 @@
-        settings.put("threadpool.bulk.queue_size", 1);
+    protected Settings nodeSettings() {
+        Settings.Builder settings = Settings.builder().put(super.nodeSettings());
+        // Use pools of size 1 so we can block them
         settings.put("threadpool.bulk.size", 1);
-        settings.put("threadpool.search.queue_size", 1);
         settings.put("threadpool.search.size", 1);
+        // Use queues of size 1 because size 0 is broken and because search requests need the queue to function
+        settings.put("threadpool.bulk.queue_size", 1);
+        settings.put("threadpool.search.queue_size", 1);
         return settings.build();
     }
 
-    /**
-     * Disable search context leak detection because we expect leaks when there is an {@link EsRejectedExecutionException} queueing the
-     * reduce phase.
-     */
-    @Override
-    protected Collection<Class<? extends Plugin>> getMockPlugins() {
-        List<Class<? extends Plugin>> mockPlugins = new ArrayList<>();
-        for (Class<? extends Plugin> plugin: super.getMockPlugins()) {
-            if (plugin.equals(MockSearchService.TestPlugin.class)) {
-                continue;
-            }
-            mockPlugins.add(plugin);
+    @Before
+    public void setupSourceIndex() throws Exception {
+        createIndex("source");
+        // Build the test data. Don't use indexRandom because that won't work consistently with such small thread pools.
+        BulkRequestBuilder bulk = client().prepareBulk();
+        for (int i = 0; i < DOC_COUNT; i++) {
+            bulk.add(client().prepareIndex("source", "test").setSource("foo", "bar " + i));
+        }
+        Retry retry = Retry.on(EsRejectedExecutionException.class).policy(BackoffPolicy.exponentialBackoff());
+        BulkResponse response = retry.withSyncBackoff(client(), bulk.request());
+        assertFalse(response.buildFailureMessage(), response.hasFailures());
+        client().admin().indices().prepareRefresh("source").get();
+    }
+
+    @After
+    public void forceUnblockAllExecutors() {
+        for (CyclicBarrier barrier: blockedExecutors) {
+            barrier.reset();
         }
-        return mockPlugins;
     }
 
     public void testReindex() throws Exception {
-        setupSourceIndex("source");
-        testCase(true, i -> reindex().source("source").destination("dest" + i));
+        testCase(ReindexAction.NAME, ReindexAction.INSTANCE.newRequestBuilder(client()).source("source").destination("dest"),
+                matcher().created(DOC_COUNT));
     }
 
     public void testUpdateByQuery() throws Exception {
-        for (int i = 0; i < CONCURRENT; i++) {
-            setupSourceIndex("source" + i);
-        }
-        testCase(false, i -> updateByQuery().source("source" + i));
+        testCase(UpdateByQueryAction.NAME, UpdateByQueryAction.INSTANCE.newRequestBuilder(client()).source("source"),
+                matcher().updated(DOC_COUNT));
     }
```

Comment:
Nik, sorry to sneak in but do you think you can add a test for Delete-By-Query here?